packages/monitor-v2/src/monitor-og/SnapshotVerification.ts

```diff
@@ -0,0 +1,333 @@
+    ipfsProposal.plugins.safeSnap === undefined ||
+    graphqlProposal.plugins.safeSnap === undefined ||
+    ipfsProposal.plugins.safeSnap.safes.length !== graphqlProposal.plugins.safeSnap.safes.length
+  ) {
+    return false;
+  }
+
+  // Verify that the safes match. We inspect only the properties that are required for verification.
+  for (let i = 0; i < ipfsProposal.plugins.safeSnap.safes.length; i++) {
+    const ipfsSafe = ipfsProposal.plugins.safeSnap.safes[i];
+    const graphqlSafe = graphqlProposal.plugins.safeSnap.safes[i];
+
+    // Verify the network, oSnap address and number of transaction batches.
+    if (
+      ipfsSafe.network !== graphqlSafe.network ||
+      ipfsSafe.umaAddress !== graphqlSafe.umaAddress ||
+      ipfsSafe.txs.length !== graphqlSafe.txs.length
+    ) {
+      return false;
+    }
+
+    // Verify that the transaction batch properties match.
+    for (let j = 0; j < ipfsSafe.txs.length; j++) {
+      const ipfsTransaction = ipfsSafe.txs[j].mainTransaction;
+      const graphqlTransaction = graphqlSafe.txs[j].mainTransaction;
+
+      if (
+        ipfsTransaction.to !== graphqlTransaction.to ||
+        ipfsTransaction.data !== graphqlTransaction.data ||
+        ipfsTransaction.value !== graphqlTransaction.value ||
+        ipfsTransaction.operation !== graphqlTransaction.operation
+      ) {
+        return false;
+      }
+    }
+  }
+
+  return true;
+};
+
+const findSafe = (safe: SafeSnapSafe, chainId: number, ogAddress: string): boolean => {
+  return (
+    safe.network === chainId.toString() && ethersUtils.getAddress(safe.umaAddress) === ethersUtils.getAddress(ogAddress)
+  );
+};
+
+export const verifyProposal = async (
+  transaction: TransactionsProposedEvent,
+  params: MonitoringParams
+): Promise<VerificationResponse> => {
+  const ipfsHash = tryHexToUtf8String(transaction.args.explanation);
+  // tryHexToUtf8String returns the input if it is not decoding to UTF-8 string.
+  if (ipfsHash === transaction.args.explanation) {
+    return { verified: false, error: `Could not decode explanation ${transaction.args.explanation}` };
+  }
+
+  // Get proposal data from GraphQL.
+  const graphqlData = await getGraphqlData(ipfsHash, params.graphqlEndpoint);
+  if (graphqlData instanceof Error) {
+    return { verified: false, error: `GraphQL request failed with error ${graphqlData.message}` };
+  }
+
+  // Verify that the proposal exists and is unique.
+  if (graphqlData.proposals.length === 0) {
+    return { verified: false, error: `No proposal found for IPFS hash ${ipfsHash}` };
+  } else if (graphqlData.proposals.length > 1) {
+    return { verified: false, error: `Duplicate proposals found for IPFS hash ${ipfsHash}` };
+  }
+
+  // Verify proposal type.
+  const proposal = graphqlData.proposals[0];
+  if (proposal.type !== "single-choice" && proposal.type !== "basic") {
+    return { verified: false, error: `Proposal type ${proposal.type} is not supported` };
+  }
+  if (proposal.type === "basic") {
```

Comment:
It seems that we are not checking the proposal when `type==="single-choice"`. Not sure if it's necessary though.

Reply:
We check that the basic proposal has expected choices. This is done through assertion as this should always evaluate to true unless the snapshot changed the hardcoded choices for basic proposals. For single choice we do not have similar check as proposer could enter any number and labels for choices that we try to verify by matching them to known approval choice labels.

Reply:
Sorry I'm a bit confused -- where do we evaluate these single-choice votes?

Reply:
To avoid confusion I moved approval label matching right after basic type assertion. And applied approval label matching only to single choice as basic should have known fixed choices.